dags/lib/tasks/batch_type.py

```diff
@@ -46,10 +54,108 @@ def _validate_cnv_vcf_files(metadata: dict, cnv_suffix: str):
+
+    Here batch_id and sequencing_ids are mutually exclusive. If both are provided, an exception will be raised.
+
+    If a `batch_id` is provided and the analysis type cannot be determined from the `enriched_clinical` table,
+    the function will attempt to infer the analysis type from the metadata file. If the metadata file does
+    not exist, the analysis type will default to `SOMATIC_TUMOR_NORMAL`.
+
+    The possible analysis types (formerly referred to as batch type) are:
+        - GERMLINE
+        - SOMATIC_TUMOR_ONLY
+        - SOMATIC_TUMOR_NORMAL
+    """
+    import logging
+    from airflow.exceptions import AirflowFailException
+    from lib.tasks.batch_type import _detect_type_from_enrich_clinical, _detect_type_from_metadata_file
+
+    logger = logging.getLogger(__name__)
+
+    if batch_id and sequencing_ids:
+        raise AirflowFailException("Only one of batch_id or sequencing_ids can be provided")
+
+    if not (batch_id or sequencing_ids):
+        raise AirflowFailException("Either batch_id or sequencing_ids must be provided")
+
+    identifier_to_type = _detect_type_from_enrich_clinical(
+        identifier_column="batch_id" if batch_id else "service_request_id",
+        identifiers=[batch_id] if batch_id else sequencing_ids,
+        must_exist=bool(sequencing_ids)
+    )
+    if not identifier_to_type and batch_id:
+        logger.info(f"Unable to infer batch type for batch ID {batch_id} from the enriched clinical table. Falling back to metadata file.")
+        return _detect_type_from_metadata_file(batch_id)
+    else:
+        return identifier_to_type
+
+
+def _detect_type_from_enrich_clinical(identifier_column, identifiers, must_exist=True):
+    """
+    Returns a dictionary mapping each identifier to its corresponding analysis type.
+
+    The possible analysis types (formerly referred to as batch type) are:
+     - GERMLINE
+     - SOMATIC_TUMOR_ONLY
+     - SOMATIC_TUMOR_NORMAL
+
+    The identifiers provided must match the values in the specified `identifier_column`
+    of the `enriched_clinical` table.
+    """
+    from collections import defaultdict
+    from airflow.exceptions import AirflowFailException
+    from pandas import DataFrame
+    from lib.datasets import enriched_clinical
+    from lib.utils_etl_tables import to_pandas
+
+    code_to_type: dict = {
+        BioinfoAnalysisCode.GEBA.value: ClinAnalysis.GERMLINE.value,
+        BioinfoAnalysisCode.TEBA.value: ClinAnalysis.SOMATIC_TUMOR_ONLY.value,
+        BioinfoAnalysisCode.TNEBA.value:  ClinAnalysis.SOMATIC_TUMOR_NORMAL.value
+    }
+
+    df: DataFrame = (
+        to_pandas(enriched_clinical.uri)
+        .filter([identifier_column, "bioinfo_analysis_code"])
+        .set_index(identifier_column)
+    )
+    distinct_pairs_df = df[df.index.isin(identifiers)].drop_duplicates()
+
+    identifier_to_codes = defaultdict(list)
+    for tup in distinct_pairs_df.itertuples():
+        identifier_to_codes[tup[0]].append(tup[1])
```

Comment:
Nommer tes variables locales :
```suggestion
    for _id, codes in distinct_pairs_df.itertuples():
        identifier_to_codes[_id].append(codes)
```

Reply:
C'est fait